File: performance_validation/benchmarks/real_implementations/parser_bench.cpp
// Real Parser Benchmark - C++ Implementation
#include <iostream>
#include <fstream>
#include <sstream>
#include <chrono>
#include <vector>
#include <string>
#include <stack>
#include <unordered_map>
#include <memory>

// Simple AST node structure
struct ASTNode {
    std::string type;
    std::string value;
    std::vector<std::shared_ptr<ASTNode>> children;
    
    ASTNode(const std::string& t, const std::string& v = "") : type(t), value(v) {}
};

class SimpleParser {
private:
    std::string input;
    size_t position;
    int nodesCreated;
    
    // Token types
    enum TokenType {
        IDENTIFIER, NUMBER, STRING, KEYWORD, OPERATOR, DELIMITER, END
    };
    
    struct Token {
        TokenType type;
        std::string value;
    };
    
    Token currentToken;
    
    void skipWhitespace() {
        while (position < input.length() && std::isspace(input[position])) {
            position++;
        }
    }
    
    Token nextToken() {
        skipWhitespace();
        
        if (position >= input.length()) {
            return {END, ""};
        }
        
        char c = input[position];
        
        // Identifier or keyword
        if (std::isalpha(c) || c == '_') {
            std::string value;
            while (position < input.length() && (std::isalnum(input[position]) || input[position] == '_')) {
                value += input[position++];
            }
            
            // Check if it's a keyword
            static const std::unordered_map<std::string, bool> keywords = {
                {"fun", true}, {"val", true}, {"var", true}, {"if", true},
                {"else", true}, {"while", true}, {"for", true}, {"return", true},
                {"class", true}, {"interface", true}, {"import", true}
            };
            
            return {keywords.count(value) ? KEYWORD : IDENTIFIER, value};
        }
        
        // Number
        if (std::isdigit(c)) {
            std::string value;
            while (position < input.length() && (std::isdigit(input[position]) || input[position] == '.')) {
                value += input[position++];
            }
            return {NUMBER, value};
        }
        
        // String
        if (c == '"' || c == '\'') {
            char quote = c;
            position++;
            std::string value;
            while (position < input.length() && input[position] != quote) {
                if (input[position] == '\\' && position + 1 < input.length()) {
                    position++; // Skip escape sequence
                }
                value += input[position++];
            }
            if (position < input.length()) position++; // Skip closing quote
            return {STRING, value};
        }
        
        // Operators and delimiters
        static const std::string operators = "+-*/%=<>!&|";
        static const std::string delimiters = "(){}[],;:.";
        
        if (operators.find(c) != std::string::npos) {
            std::string value(1, c);
            position++;
            // Check for two-character operators
            if (position < input.length()) {
                char next = input[position];
                if ((c == '=' && next == '=') || (c == '!' && next == '=') ||
                    (c == '<' && next == '=') || (c == '>' && next == '=') ||
                    (c == '&' && next == '&') || (c == '|' && next == '|')) {
                    value += next;
                    position++;
                }
            }
            return {OPERATOR, value};
        }
        
        if (delimiters.find(c) != std::string::npos) {
            position++;
            return {DELIMITER, std::string(1, c)};
        }
        
        // Unknown character, skip it
        position++;
        return nextToken();
    }
    
public:
    SimpleParser(const std::string& code) : input(code), position(0), nodesCreated(0) {}
    
    std::shared_ptr<ASTNode> parse() {
        auto root = std::make_shared<ASTNode>("Program");
        nodesCreated++;
        
        currentToken = nextToken();
        
        while (currentToken.type != END) {
            auto stmt = parseStatement();
            if (stmt) {
                root->children.push_back(stmt);
            }
        }
        
        return root;
    }
    
    std::shared_ptr<ASTNode> parseStatement() {
        if (currentToken.type == KEYWORD) {
            if (currentToken.value == "fun") {
                return parseFunctionDeclaration();
            } else if (currentToken.value == "val" || currentToken.value == "var") {
                return parseVariableDeclaration();
            } else if (currentToken.value == "if") {
                return parseIfStatement();
            } else if (currentToken.value == "while") {
                return parseWhileStatement();
            } else if (currentToken.value == "for") {
                return parseForStatement();
            } else if (currentToken.value == "return") {
                return parseReturnStatement();
            } else if (currentToken.value == "class") {
                return parseClassDeclaration();
            }
        }
        
        return parseExpression();
    }
    
    std::shared_ptr<ASTNode> parseFunctionDeclaration() {
        auto node = std::make_shared<ASTNode>("FunctionDecl");
        nodesCreated++;
        
        currentToken = nextToken(); // Skip 'fun'
        if (currentToken.type == IDENTIFIER) {
            node->value = currentToken.value;
            currentToken = nextToken();
        }
        
        // Parse parameters
        if (currentToken.type == DELIMITER && currentToken.value == "(") {
            currentToken = nextToken();
            while (currentToken.type != DELIMITER || currentToken.value != ")") {
                if (currentToken.type == IDENTIFIER) {
                    auto param = std::make_shared<ASTNode>("Parameter", currentToken.value);
                    nodesCreated++;
                    node->children.push_back(param);
                }
                currentToken = nextToken();
            }
            currentToken = nextToken(); // Skip ')'
        }
        
        // Parse body
        if (currentToken.type == DELIMITER && currentToken.value == "{") {
            node->children.push_back(parseBlock());
        }
        
        return node;
    }
    
    std::shared_ptr<ASTNode> parseVariableDeclaration() {
        auto node = std::make_shared<ASTNode>("VarDecl", currentToken.value);
        nodesCreated++;
        
        currentToken = nextToken(); // Skip 'val' or 'var'
        if (currentToken.type == IDENTIFIER) {
            auto id = std::make_shared<ASTNode>("Identifier", currentToken.value);
            nodesCreated++;
            node->children.push_back(id);
            currentToken = nextToken();
        }
        
        if (currentToken.type == OPERATOR && currentToken.value == "=") {
            currentToken = nextToken();
            node->children.push_back(parseExpression());
        }
        
        return node;
    }
    
    std::shared_ptr<ASTNode> parseIfStatement() {
        auto node = std::make_shared<ASTNode>("IfStatement");
        nodesCreated++;
        
        currentToken = nextToken(); // Skip 'if'
        if (currentToken.type == DELIMITER && currentToken.value == "(") {
            currentToken = nextToken();
            node->children.push_back(parseExpression());
            if (currentToken.type == DELIMITER && currentToken.value == ")") {
                currentToken = nextToken();
            }
        }
        
        node->children.push_back(parseStatement());
        
        if (currentToken.type == KEYWORD && currentToken.value == "else") {
            currentToken = nextToken();
            node->children.push_back(parseStatement());
        }
        
        return node;
    }
    
    std::shared_ptr<ASTNode> parseWhileStatement() {
        auto node = std::make_shared<ASTNode>("WhileStatement");
        nodesCreated++;
        
        currentToken = nextToken(); // Skip 'while'
        if (currentToken.type == DELIMITER && currentToken.value == "(") {
            currentToken = nextToken();
            node->children.push_back(parseExpression());
            if (currentToken.type == DELIMITER && currentToken.value == ")") {
                currentToken = nextToken();
            }
        }
        
        node->children.push_back(parseStatement());
        return node;
    }
    
    std::shared_ptr<ASTNode> parseForStatement() {
        auto node = std::make_shared<ASTNode>("ForStatement");
        nodesCreated++;
        
        currentToken = nextToken(); // Skip 'for'
        if (currentToken.type == DELIMITER && currentToken.value == "(") {
            currentToken = nextToken();
            // Simple for loop parsing
            while (currentToken.type != DELIMITER || currentToken.value != ")") {
                currentToken = nextToken();
            }
            currentToken = nextToken(); // Skip ')'
        }
        
        node->children.push_back(parseStatement());
        return node;
    }
    
    std::shared_ptr<ASTNode> parseReturnStatement() {
        auto node = std::make_shared<ASTNode>("ReturnStatement");
        nodesCreated++;
        
        currentToken = nextToken(); // Skip 'return'
        if (currentToken.type != DELIMITER || currentToken.value != ";") {
            node->children.push_back(parseExpression());
        }
        
        return node;
    }
    
    std::shared_ptr<ASTNode> parseClassDeclaration() {
        auto node = std::make_shared<ASTNode>("ClassDecl");
        nodesCreated++;
        
        currentToken = nextToken(); // Skip 'class'
        if (currentToken.type == IDENTIFIER) {
            node->value = currentToken.value;
            currentToken = nextToken();
        }
        
        if (currentToken.type == DELIMITER && currentToken.value == "{") {
            node->children.push_back(parseBlock());
        }
        
        return node;
    }
    
    std::shared_ptr<ASTNode> parseBlock() {
        auto node = std::make_shared<ASTNode>("Block");
        nodesCreated++;
        
        if (currentToken.type == DELIMITER && currentToken.value == "{") {
            currentToken = nextToken();
            while (currentToken.type != DELIMITER || currentToken.value != "}") {
                auto stmt = parseStatement();
                if (stmt) {
                    node->children.push_back(stmt);
                }
                if (currentToken.type == DELIMITER && currentToken.value == ";") {
                    currentToken = nextToken();
                }
            }
            currentToken = nextToken(); // Skip '}'
        }
        
        return node;
    }
    
    std::shared_ptr<ASTNode> parseExpression() {
        return parseBinaryExpression(0);
    }
    
    std::shared_ptr<ASTNode> parseBinaryExpression(int minPrecedence) {
        auto left = parsePrimaryExpression();
        
        while (currentToken.type == OPERATOR) {
            std::string op = currentToken.value;
            int precedence = getOperatorPrecedence(op);
            
            if (precedence < minPrecedence) {
                break;
            }
            
            currentToken = nextToken();
            auto right = parseBinaryExpression(precedence + 1);
            
            auto binOp = std::make_shared<ASTNode>("BinaryOp", op);
            nodesCreated++;
            binOp->children.push_back(left);
            binOp->children.push_back(right);
            left = binOp;
        }
        
        return left;
    }
    
    std::shared_ptr<ASTNode> parsePrimaryExpression() {
        if (currentToken.type == NUMBER) {
            auto node = std::make_shared<ASTNode>("Number", currentToken.value);
            nodesCreated++;
            currentToken = nextToken();
            return node;
        }
        
        if (currentToken.type == STRING) {
            auto node = std::make_shared<ASTNode>("String", currentToken.value);
            nodesCreated++;
            currentToken = nextToken();
            return node;
        }
        
        if (currentToken.type == IDENTIFIER) {
            auto node = std::make_shared<ASTNode>("Identifier", currentToken.value);
            nodesCreated++;
            currentToken = nextToken();
            
            // Check for function call
            if (currentToken.type == DELIMITER && currentToken.value == "(") {
                auto call = std::make_shared<ASTNode>("FunctionCall");
                nodesCreated++;
                call->children.push_back(node);
                
                currentToken = nextToken();
                while (currentToken.type != DELIMITER || currentToken.value != ")") {
                    call->children.push_back(parseExpression());
                    if (currentToken.type == DELIMITER && currentToken.value == ",") {
                        currentToken = nextToken();
                    }
                }
                currentToken = nextToken(); // Skip ')'
                return call;
            }
            
            return node;
        }
        
        if (currentToken.type == DELIMITER && currentToken.value == "(") {
            currentToken = nextToken();
            auto expr = parseExpression();
            if (currentToken.type == DELIMITER && currentToken.value == ")") {
                currentToken = nextToken();
            }
            return expr;
        }
        
        // Skip unknown tokens
        currentToken = nextToken();
        return nullptr;
    }
    
    int getOperatorPrecedence(const std::string& op) {
        static const std::unordered_map<std::string, int> precedence = {
            {"=", 1},
            {"||", 2},
            {"&&", 3},
            {"==", 4}, {"!=", 4},
            {"<", 5}, {">", 5}, {"<=", 5}, {">=", 5},
            {"+", 6}, {"-", 6},
            {"*", 7}, {"/", 7}, {"%", 7}
        };
        
        auto it = precedence.find(op);
        return it != precedence.end() ? it->second : 0;
    }
    
    int getNodeCount() const { return nodesCreated; }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <input_file> [iterations]" << std::endl;
        return 1;
    }
    
    // Read input file
    std::ifstream file(argv[1]);
    if (!file) {
        std::cerr << "Error: Cannot open file " << argv[1] << std::endl;
        return 1;
    }
    
    std::stringstream buffer;
    buffer << file.rdbuf();
    std::string input = buffer.str();
    
    int iterations = argc > 2 ? std::atoi(argv[2]) : 30;
    
    // Warm-up
    for (int i = 0; i < 5; i++) {
        SimpleParser warmup(input);
        warmup.parse();
    }
    
    // Benchmark
    std::vector<double> times;
    int totalNodes = 0;
    
    for (int i = 0; i < iterations; i++) {
        auto start = std::chrono::high_resolution_clock::now();
        
        SimpleParser parser(input);
        auto ast = parser.parse();
        
        auto end = std::chrono::high_resolution_clock::now();
        double elapsed = std::chrono::duration<double>(end - start).count();
        
        times.push_back(elapsed);
        totalNodes = parser.getNodeCount();
    }
    
    // Calculate statistics
    double sum = 0;
    for (double t : times) sum += t;
    double mean = sum / times.size();
    
    // Output JSON results
    std::cout << "{\n";
    std::cout << "  \"language\": \"cpp\",\n";
    std::cout << "  \"benchmark\": \"parser\",\n";
    std::cout << "  \"iterations\": " << iterations << ",\n";
    std::cout << "  \"nodes_processed\": " << totalNodes << ",\n";
    std::cout << "  \"times\": [";
    for (size_t i = 0; i < times.size(); i++) {
        if (i > 0) std::cout << ", ";
        std::cout << times[i];
    }
    std::cout << "],\n";
    std::cout << "  \"average_time\": " << mean << ",\n";
    std::cout << "  \"nodes_per_second\": " << (totalNodes / mean) << "\n";
    std::cout << "}\n";
    
    return 0;
}

File: performance_validation/benchmarks/real_implementations/parser_bench.rs
// Real Parser Benchmark - Rust Implementation
use std::env;
use std::fs;
use std::time::Instant;

#[derive(Debug, Clone)]
enum TokenType {
    Identifier,
    Number,
    String,
    Keyword,
    Operator,
    Delimiter,
    End,
}

#[derive(Debug, Clone)]
struct Token {
    token_type: TokenType,
    value: String,
}

#[derive(Debug)]
struct ASTNode {
    node_type: String,
    value: String,
    children: Vec<Box<ASTNode>>,
}

impl ASTNode {
    fn new(node_type: &str) -> Self {
        ASTNode {
            node_type: node_type.to_string(),
            value: String::new(),
            children: Vec::new(),
        }
    }
    
    fn new_with_value(node_type: &str, value: &str) -> Self {
        ASTNode {
            node_type: node_type.to_string(),
            value: value.to_string(),
            children: Vec::new(),
        }
    }
}

struct SimpleParser {
    input: Vec<char>,
    position: usize,
    current_token: Token,
    nodes_created: usize,
}

impl SimpleParser {
    fn new(input: &str) -> Self {
        let mut parser = SimpleParser {
            input: input.chars().collect(),
            position: 0,
            current_token: Token { token_type: TokenType::End, value: String::new() },
            nodes_created: 0,
        };
        parser.current_token = parser.next_token();
        parser
    }
    
    fn skip_whitespace(&mut self) {
        while self.position < self.input.len() && self.input[self.position].is_whitespace() {
            self.position += 1;
        }
    }
    
    fn next_token(&mut self) -> Token {
        self.skip_whitespace();
        
        if self.position >= self.input.len() {
            return Token { token_type: TokenType::End, value: String::new() };
        }
        
        let c = self.input[self.position];
        
        // Identifier or keyword
        if c.is_alphabetic() || c == '_' {
            let mut value = String::new();
            while self.position < self.input.len() && 
                  (self.input[self.position].is_alphanumeric() || self.input[self.position] == '_') {
                value.push(self.input[self.position]);
                self.position += 1;
            }
            
            let token_type = match value.as_str() {
                "fun" | "val" | "var" | "if" | "else" | "while" | "for" | 
                "return" | "class" | "interface" | "import" => TokenType::Keyword,
                _ => TokenType::Identifier,
            };
            
            return Token { token_type, value };
        }
        
        // Number
        if c.is_numeric() {
            let mut value = String::new();
            while self.position < self.input.len() && 
                  (self.input[self.position].is_numeric() || self.input[self.position] == '.') {
                value.push(self.input[self.position]);
                self.position += 1;
            }
            return Token { token_type: TokenType::Number, value };
        }
        
        // String
        if c == '"' || c == '\'' {
            let quote = c;
            self.position += 1;
            let mut value = String::new();
            while self.position < self.input.len() && self.input[self.position] != quote {
                if self.input[self.position] == '\\' && self.position + 1 < self.input.len() {
                    self.position += 1;
                }
                value.push(self.input[self.position]);
                self.position += 1;
            }
            if self.position < self.input.len() {
                self.position += 1; // Skip closing quote
            }
            return Token { token_type: TokenType::String, value };
        }
        
        // Operators and delimiters
        let operators = "+-*/%=<>!&|";
        let delimiters = "(){}[],;:.";
        
        if operators.contains(c) {
            let mut value = String::from(c);
            self.position += 1;
            
            // Check for two-character operators
            if self.position < self.input.len() {
                let next = self.input[self.position];
                if (c == '=' && next == '=') || (c == '!' && next == '=') ||
                   (c == '<' && next == '=') || (c == '>' && next == '=') ||
                   (c == '&' && next == '&') || (c == '|' && next == '|') {
                    value.push(next);
                    self.position += 1;
                }
            }
            return Token { token_type: TokenType::Operator, value };
        }
        
        if delimiters.contains(c) {
            self.position += 1;
            return Token { token_type: TokenType::Delimiter, value: String::from(c) };
        }
        
        // Unknown character, skip it
        self.position += 1;
        self.next_token()
    }
    
    fn parse(&mut self) -> Box<ASTNode> {
        let mut root = Box::new(ASTNode::new("Program"));
        self.nodes_created += 1;
        
        while !matches!(self.current_token.token_type, TokenType::End) {
            if let Some(stmt) = self.parse_statement() {
                root.children.push(stmt);
            }
        }
        
        root
    }
    
    fn parse_statement(&mut self) -> Option<Box<ASTNode>> {
        match &self.current_token.token_type {
            TokenType::Keyword => {
                match self.current_token.value.as_str() {
                    "fun" => Some(self.parse_function_declaration()),
                    "val" | "var" => Some(self.parse_variable_declaration()),
                    "if" => Some(self.parse_if_statement()),
                    "while" => Some(self.parse_while_statement()),
                    "for" => Some(self.parse_for_statement()),
                    "return" => Some(self.parse_return_statement()),
                    "class" => Some(self.parse_class_declaration()),
                    _ => self.parse_expression(),
                }
            }
            _ => self.parse_expression(),
        }
    }
    
    fn parse_function_declaration(&mut self) -> Box<ASTNode> {
        let mut node = Box::new(ASTNode::new("FunctionDecl"));
        self.nodes_created += 1;
        
        self.current_token = self.next_token(); // Skip 'fun'
        if matches!(self.current_token.token_type, TokenType::Identifier) {
            node.value = self.current_token.value.clone();
            self.current_token = self.next_token();
        }
        
        // Parse parameters
        if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == "(" {
            self.current_token = self.next_token();
            while !(matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == ")") {
                if matches!(self.current_token.token_type, TokenType::Identifier) {
                    let param = Box::new(ASTNode::new_with_value("Parameter", &self.current_token.value));
                    self.nodes_created += 1;
                    node.children.push(param);
                }
                self.current_token = self.next_token();
            }
            self.current_token = self.next_token(); // Skip ')'
        }
        
        // Parse body
        if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == "{" {
            node.children.push(self.parse_block());
        }
        
        node
    }
    
    fn parse_variable_declaration(&mut self) -> Box<ASTNode> {
        let mut node = Box::new(ASTNode::new_with_value("VarDecl", &self.current_token.value));
        self.nodes_created += 1;
        
        self.current_token = self.next_token(); // Skip 'val' or 'var'
        if matches!(self.current_token.token_type, TokenType::Identifier) {
            let id = Box::new(ASTNode::new_with_value("Identifier", &self.current_token.value));
            self.nodes_created += 1;
            node.children.push(id);
            self.current_token = self.next_token();
        }
        
        if matches!(self.current_token.token_type, TokenType::Operator) && self.current_token.value == "=" {
            self.current_token = self.next_token();
            if let Some(expr) = self.parse_expression() {
                node.children.push(expr);
            }
        }
        
        node
    }
    
    fn parse_if_statement(&mut self) -> Box<ASTNode> {
        let mut node = Box::new(ASTNode::new("IfStatement"));
        self.nodes_created += 1;
        
        self.current_token = self.next_token(); // Skip 'if'
        if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == "(" {
            self.current_token = self.next_token();
            if let Some(expr) = self.parse_expression() {
                node.children.push(expr);
            }
            if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == ")" {
                self.current_token = self.next_token();
            }
        }
        
        if let Some(stmt) = self.parse_statement() {
            node.children.push(stmt);
        }
        
        if matches!(self.current_token.token_type, TokenType::Keyword) && self.current_token.value == "else" {
            self.current_token = self.next_token();
            if let Some(stmt) = self.parse_statement() {
                node.children.push(stmt);
            }
        }
        
        node
    }
    
    fn parse_while_statement(&mut self) -> Box<ASTNode> {
        let mut node = Box::new(ASTNode::new("WhileStatement"));
        self.nodes_created += 1;
        
        self.current_token = self.next_token(); // Skip 'while'
        if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == "(" {
            self.current_token = self.next_token();
            if let Some(expr) = self.parse_expression() {
                node.children.push(expr);
            }
            if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == ")" {
                self.current_token = self.next_token();
            }
        }
        
        if let Some(stmt) = self.parse_statement() {
            node.children.push(stmt);
        }
        
        node
    }
    
    fn parse_for_statement(&mut self) -> Box<ASTNode> {
        let mut node = Box::new(ASTNode::new("ForStatement"));
        self.nodes_created += 1;
        
        self.current_token = self.next_token(); // Skip 'for'
        if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == "(" {
            self.current_token = self.next_token();
            while !(matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == ")") {
                self.current_token = self.next_token();
            }
            self.current_token = self.next_token(); // Skip ')'
        }
        
        if let Some(stmt) = self.parse_statement() {
            node.children.push(stmt);
        }
        
        node
    }
    
    fn parse_return_statement(&mut self) -> Box<ASTNode> {
        let mut node = Box::new(ASTNode::new("ReturnStatement"));
        self.nodes_created += 1;
        
        self.current_token = self.next_token(); // Skip 'return'
        if !(matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == ";") {
            if let Some(expr) = self.parse_expression() {
                node.children.push(expr);
            }
        }
        
        node
    }
    
    fn parse_class_declaration(&mut self) -> Box<ASTNode> {
        let mut node = Box::new(ASTNode::new("ClassDecl"));
        self.nodes_created += 1;
        
        self.current_token = self.next_token(); // Skip 'class'
        if matches!(self.current_token.token_type, TokenType::Identifier) {
            node.value = self.current_token.value.clone();
            self.current_token = self.next_token();
        }
        
        if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == "{" {
            node.children.push(self.parse_block());
        }
        
        node
    }
    
    fn parse_block(&mut self) -> Box<ASTNode> {
        let mut node = Box::new(ASTNode::new("Block"));
        self.nodes_created += 1;
        
        if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == "{" {
            self.current_token = self.next_token();
            while !(matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == "}") {
                if let Some(stmt) = self.parse_statement() {
                    node.children.push(stmt);
                }
                if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == ";" {
                    self.current_token = self.next_token();
                }
                if matches!(self.current_token.token_type, TokenType::End) {
                    break;
                }
            }
            if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == "}" {
                self.current_token = self.next_token(); // Skip '}'
            }
        }
        
        node
    }
    
    fn parse_expression(&mut self) -> Option<Box<ASTNode>> {
        self.parse_binary_expression(0)
    }
    
    fn parse_binary_expression(&mut self, min_precedence: i32) -> Option<Box<ASTNode>> {
        let mut left = self.parse_primary_expression()?;
        
        while matches!(self.current_token.token_type, TokenType::Operator) {
            let op = self.current_token.value.clone();
            let precedence = self.get_operator_precedence(&op);
            
            if precedence < min_precedence {
                break;
            }
            
            self.current_token = self.next_token();
            let right = self.parse_binary_expression(precedence + 1)?;
            
            let mut bin_op = Box::new(ASTNode::new_with_value("BinaryOp", &op));
            self.nodes_created += 1;
            bin_op.children.push(left);
            bin_op.children.push(right);
            left = bin_op;
        }
        
        Some(left)
    }
    
    fn parse_primary_expression(&mut self) -> Option<Box<ASTNode>> {
        match &self.current_token.token_type {
            TokenType::Number => {
                let node = Box::new(ASTNode::new_with_value("Number", &self.current_token.value));
                self.nodes_created += 1;
                self.current_token = self.next_token();
                Some(node)
            }
            TokenType::String => {
                let node = Box::new(ASTNode::new_with_value("String", &self.current_token.value));
                self.nodes_created += 1;
                self.current_token = self.next_token();
                Some(node)
            }
            TokenType::Identifier => {
                let node = Box::new(ASTNode::new_with_value("Identifier", &self.current_token.value));
                self.nodes_created += 1;
                self.current_token = self.next_token();
                
                // Check for function call
                if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == "(" {
                    let mut call = Box::new(ASTNode::new("FunctionCall"));
                    self.nodes_created += 1;
                    call.children.push(node);
                    
                    self.current_token = self.next_token();
                    while !(matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == ")") {
                        if let Some(expr) = self.parse_expression() {
                            call.children.push(expr);
                        }
                        if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == "," {
                            self.current_token = self.next_token();
                        }
                        if matches!(self.current_token.token_type, TokenType::End) {
                            break;
                        }
                    }
                    if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == ")" {
                        self.current_token = self.next_token(); // Skip ')'
                    }
                    Some(call)
                } else {
                    Some(node)
                }
            }
            TokenType::Delimiter if self.current_token.value == "(" => {
                self.current_token = self.next_token();
                let expr = self.parse_expression();
                if matches!(self.current_token.token_type, TokenType::Delimiter) && self.current_token.value == ")" {
                    self.current_token = self.next_token();
                }
                expr
            }
            _ => {
                // Skip unknown tokens
                self.current_token = self.next_token();
                None
            }
        }
    }
    
    fn get_operator_precedence(&self, op: &str) -> i32 {
        match op {
            "=" => 1,
            "||" => 2,
            "&&" => 3,
            "==" | "!=" => 4,
            "<" | ">" | "<=" | ">=" => 5,
            "+" | "-" => 6,
            "*" | "/" | "%" => 7,
            _ => 0,
        }
    }
    
    fn get_node_count(&self) -> usize {
        self.nodes_created
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <input_file> [iterations]", args[0]);
        std::process::exit(1);
    }
    
    // Read input file
    let input = fs::read_to_string(&args[1])
        .expect(&format!("Error: Cannot read file {}", args[1]));
    
    let iterations = if args.len() > 2 {
        args[2].parse().unwrap_or(30)
    } else {
        30
    };
    
    // Warm-up
    for _ in 0..5 {
        let mut parser = SimpleParser::new(&input);
        parser.parse();
    }
    
    // Benchmark
    let mut times = Vec::new();
    let mut total_nodes = 0;
    
    for _ in 0..iterations {
        let start = Instant::now();
        
        let mut parser = SimpleParser::new(&input);
        let _ast = parser.parse();
        
        let elapsed = start.elapsed().as_secs_f64();
        times.push(elapsed);
        total_nodes = parser.get_node_count();
    }
    
    // Calculate statistics
    let sum: f64 = times.iter().sum();
    let mean = sum / times.len() as f64;
    
    // Output JSON results
    println!("{{");
    println!("  \"language\": \"rust\",");
    println!("  \"benchmark\": \"parser\",");
    println!("  \"iterations\": {},", iterations);
    println!("  \"nodes_processed\": {},", total_nodes);
    print!("  \"times\": [");
    for (i, t) in times.iter().enumerate() {
        if i > 0 { print!(", "); }
        print!("{}", t);
    }
    println!("],");
    println!("  \"average_time\": {},", mean);
    println!("  \"nodes_per_second\": {}", total_nodes as f64 / mean);
    println!("}}");
}